chat-plugins/helptickets.js

```diff
@@ -41,6 +41,7 @@ let tickets = {};
 let ticketBans = {};
 
 try {
+	/** @type {typeof tickets & typeof ticketBans} */
```

Comment:
:|

Reply:
Ugh, even if I strictly define the property `banned` on `TicketState` as always `undefined`, TypeScript doesn't infer that an object of `TicketState | BannedTicketState` whose `banned` property evaluates to true _must_ be a `BannedTicketState`.

Reply:
Oh, actually, TypeScript _thinks_ that `Object.assign({}, tickets, ticketBans);` returns an object of type `{[k: string]: TicketState} & {[k: string]: BannedTicketState}`, which it thinks is equivalent to `{k: string]: TicketState & BannedTicketState}`, but `Object.assign` never actually creates intersections if the keys are the same; it just overwrites the old value with the new one instead of doing a nested assignment. I'll report that.

Reply:
> `Object.assign` never actually creates intersections if the keys are the same

That should not be reported; it's intentional. TypeScript doesn't have very powerful semantics for `Object.assign`; you just need to be aware of the caveats.

Reply:
> TypeScript doesn't infer that an object of `TicketState | BannedTicketState` whose `banned` property evaluates to true _must_ be a `BannedTicketState`.

Yeah, that one is a bug. I couldn't figure out why, but it's why I split them into two tables in the first place, because I couldn't get TypeScript to tell them apart back when they were in one table.